client/editor/editor.go

```diff
@@ -126,18 +126,142 @@ func (e *Editor) showPositions() {
 }
 
 // MoveCursor updates the Cursor position.
-func (e *Editor) MoveCursor(x, _ int) {
+func (e *Editor) MoveCursor(x, y int) {
+	// move cursor horizontally
 	newCursor := e.Cursor + x
 
-	if newCursor < 0 {
-		newCursor = 0
+	// move cursor vertically
+	if y > 0 {
+		newCursor = e.calcCursorDown()
+	}
+
+	if y < 0 {
+		newCursor = e.calcCursorUp()
 	}
+
 	if newCursor > len(e.Text) {
 		newCursor = len(e.Text)
 	}
+
+	if newCursor < 0 {
+		newCursor = 0
+	}
+
 	e.Cursor = newCursor
 }
 
+func (e *Editor) calcCursorUp() int {
+	pos := e.Cursor
+	// reset cursor if out of bounds
+	if pos > len(e.Text)-1 {
+		pos = len(e.Text) - 1
+	}
+
+	// if cursor is currently on newline, "move" it
+	if e.Text[pos] == '\n' {
+		pos--
+	}
+
+	cls := -1
+	// find the start of the line the cursor is currently on
+	for i := pos; i > 0; i-- {
+		if e.Text[i] == '\n' {
+			cls = i
+			break
+		}
+	}
+	var offset int
+	// set the cursor offset from the start of the current line
+	if cls < 0 {
+		offset = e.Cursor + 1
+	} else {
+		offset = e.Cursor - cls
+	}
+
+	pls := -1
```

Comment:
Does `pls` act like a placeholder for the previous line length?

Reply:
Resolved in https://github.com/burntcarrot/rowix/pull/24#issuecomment-1367453114.